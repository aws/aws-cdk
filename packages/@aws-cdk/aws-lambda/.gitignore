*.js
*.js.map
*.d.ts
node_modules
*.generated.ts
dist
tsconfig.json
tslint.json

.jsii

.LAST_BUILD
.nyc_output
coverage
nyc.config.js
.LAST_PACKAGE
*.snk
