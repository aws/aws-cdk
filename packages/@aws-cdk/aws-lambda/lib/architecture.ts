/**
 * Architectures supported by AWS Lambda
 */
export class Architecture {
  /**
   * 64 bit architecture with x86 instruction set.
   */
  public static readonly X86_64 = new Architecture('x86_64', 'linux/amd64');

  /**
   * 64 bit architecture with the ARM instruction set.
   */
  public static readonly ARM_64 = new Architecture('arm64', 'linux/arm64');

  /**
   * Used to specify a custom architecture name.
   * Use this if the architecture name is not yet supported by the CDK.
   * @param name the architecture name as recognized by AWS Lambda.
   * @param [dockerPlatform=linux/amd64] the platform to use for this architecture when building with Docker
   */
  public static custom(name: string, dockerPlatform?: string) {
    return new Architecture(name, dockerPlatform ?? 'linux/amd64');
  }

  /**
   * The name of the architecture as recognized by the AWS Lambda service APIs.
   */
  public readonly name: string;

  /**
   * The platform to use for this architecture when building with Docker.
   */
  public readonly dockerPlatform: string;

  private constructor(archName: string, dockerPlatform: string) {
    this.name = archName;
    this.dockerPlatform = dockerPlatform;
  }
}
