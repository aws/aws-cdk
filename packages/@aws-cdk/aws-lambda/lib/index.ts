export * from './alias';
export * from './dlq';
export * from './function-base';
export * from './function';
export * from './handler';
export * from './image-function';
export * from './layers';
export * from './permission';
export * from './runtime';
export * from './code';
export * from './filesystem';
export * from './lambda-version';
export * from './singleton-lambda';
export * from './event-source';
export * from './event-source-mapping';
export * from './destination';
export * from './event-invoke-config';
export * from './scalable-attribute-api';
export * from './code-signing-config';
export * from './lambda-insights';
export * from './log-retention';

// AWS::Lambda CloudFormation Resources:
export * from './lambda.generated';

import './lambda-augmentations.generated';
