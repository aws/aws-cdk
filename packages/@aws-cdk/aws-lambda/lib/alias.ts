import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import { Construct } from '@aws-cdk/core';
import { EventInvokeConfigOptions } from './event-invoke-config';
import { IFunction, QualifiedFunctionBase } from './function-base';
import { extractQualifierFromArn, IVersion } from './lambda-version';
import { CfnAlias } from './lambda.generated';

export interface IAlias extends IFunction {
  /**
   * Name of this alias.
   *
   * @attribute
   */
  readonly aliasName: string;

  /**
   * The underlying Lambda function version.
   */
  readonly version: IVersion;
}

/**
 * Options for `lambda.Alias`.
 */
export interface AliasOptions extends EventInvokeConfigOptions {
  /**
   * Description for the alias
   *
   * @default No description
   */
  readonly description?: string;

  /**
   * Additional versions with individual weights this alias points to
   *
   * Individual additional version weights specified here should add up to
   * (less than) one. All remaining weight is routed to the default
   * version.
   *
   * For example, the config is
   *
   *    version: "1"
   *    additionalVersions: [{ version: "2", weight: 0.05 }]
   *
   * Then 5% of traffic will be routed to function version 2, while
   * the remaining 95% of traffic will be routed to function version 1.
   *
   * @default No additional versions
   */
  readonly additionalVersions?: VersionWeight[];

  /**
   * Specifies a provisioned concurrency configuration for a function's alias.
   *
   * @default No provisioned concurrency
   */
  readonly provisionedConcurrentExecutions?: number;
}

/**
 * Properties for a new Lambda alias
 */
export interface AliasProps extends AliasOptions {
  /**
   * Name of this alias
   */
  readonly aliasName: string;

  /**
   * Function version this alias refers to
   *
   * Use lambda.addVersion() to obtain a new lambda version to refer to.
   */
  readonly version: IVersion;
}

export interface AliasAttributes {
  readonly aliasName: string;
  readonly aliasVersion: IVersion;
}

/**
 * A new alias to a particular version of a Lambda function.
 */
export class Alias extends QualifiedFunctionBase implements IAlias {
  public static fromAliasAttributes(scope: Construct, id: string, attrs: AliasAttributes): IAlias {
    class Imported extends QualifiedFunctionBase implements IAlias {
      public readonly aliasName = attrs.aliasName;
      public readonly version = attrs.aliasVersion;
      public readonly lambda = attrs.aliasVersion.lambda;
      public readonly functionArn = `${attrs.aliasVersion.lambda.functionArn}:${attrs.aliasName}`;
      public readonly functionName = `${attrs.aliasVersion.lambda.functionName}:${attrs.aliasName}`;
      public readonly grantPrincipal = attrs.aliasVersion.grantPrincipal;
      public readonly role = attrs.aliasVersion.role;

      protected readonly canCreatePermissions = false;
      protected readonly qualifier = attrs.aliasName;
    }
    return new Imported(scope, id);
  }

  /**
   * Name of this alias.
   *
   * @attribute
   */
  public readonly aliasName: string;
  /**
   * ARN of this alias
   *
   * Used to be able to use Alias in place of a regular Lambda. Lambda accepts
   * ARNs everywhere it accepts function names.
   */
  public readonly functionName: string;

  public readonly lambda: IFunction;

  public readonly version: IVersion;

  /**
   * ARN of this alias
   *
   * Used to be able to use Alias in place of a regular Lambda. Lambda accepts
   * ARNs everywhere it accepts function names.
   */
  public readonly functionArn: string;

  protected readonly qualifier: string;

  protected readonly canCreatePermissions: boolean = true;

  constructor(scope: Construct, id: string, props: AliasProps) {
    super(scope, id, {
      physicalName: props.aliasName,
    });

    this.lambda = props.version.lambda;
    this.aliasName = this.physicalName;
    this.version = props.version;

    const alias = new CfnAlias(this, 'Resource', {
      name: this.aliasName,
      description: props.description,
      functionName: this.version.lambda.functionName,
      functionVersion: props.version.version,
      routingConfig: this.determineRoutingConfig(props),
      provisionedConcurrencyConfig: this.determineProvisionedConcurrency(props),
    });

    this.functionArn = this.getResourceArnAttribute(alias.ref, {
      service: 'lambda',
      resource: 'function',
      resourceName: `${this.lambda.functionName}:${this.physicalName}`,
      sep: ':',
    });

    this.qualifier = extractQualifierFromArn(alias.ref);

    if (props.onFailure || props.onSuccess || props.maxEventAge || props.retryAttempts !== undefined) {
      this.configureAsyncInvoke({
        onFailure: props.onFailure,
        onSuccess: props.onSuccess,
        maxEventAge: props.maxEventAge,
        retryAttempts: props.retryAttempts,
      });
    }

    // ARN parsing splits on `:`, so we can only get the function's name from the ARN as resourceName...
    // And we're parsing it out (instead of using the underlying function directly) in order to have use of it incur
    // an implicit dependency on the resource.
    this.functionName = `${this.stack.parseArn(this.functionArn, ':').resourceName!}:${this.aliasName}`;
  }

  public get grantPrincipal() {
    return this.version.grantPrincipal;
  }

  public get role() {
    return this.version.role;
  }

  public metric(metricName: string, props: cloudwatch.MetricOptions = {}): cloudwatch.Metric {
    // Metrics on Aliases need the "bare" function name, and the alias' ARN, this differes from the base behavior.
    return super.metric(metricName, {
      dimensions: {
        FunctionName: this.lambda.functionName,
        // construct the name from the underlying lambda so that alarms on an alias
        // don't cause a circular dependency with CodeDeploy
        // see: https://github.com/aws/aws-cdk/issues/2231
        Resource: `${this.lambda.functionName}:${this.aliasName}`,
      },
      ...props,
    });
  }

  /**
   * Calculate the routingConfig parameter from the input props
   */
  private determineRoutingConfig(props: AliasProps) {
    if (!props.additionalVersions || props.additionalVersions.length === 0) {
      return undefined;
    }

    this.validateAdditionalWeights(props.additionalVersions);

    return {
      additionalVersionWeights: props.additionalVersions.map(vw => {
        return {
          functionVersion: vw.version.version,
          functionWeight: vw.weight,
        };
      }),
    };
  }

  /**
   * Validate that the additional version weights make sense
   *
   * We validate that they are positive and add up to something <= 1.
   */
  private validateAdditionalWeights(weights: VersionWeight[]) {
    const total = weights.map(w => {
      if (w.weight < 0 || w.weight > 1) { throw new Error(`Additional version weight must be between 0 and 1, got: ${w.weight}`); }
      return w.weight;
    }).reduce((a, x) => a + x);

    if (total > 1) {
      throw new Error(`Sum of additional version weights must not exceed 1, got: ${total}`);
    }
  }

  /**
   * Validate that the provisionedConcurrentExecutions makes sense
   *
   * Member must have value greater than or equal to 1
   */
  private determineProvisionedConcurrency(props: AliasProps): CfnAlias.ProvisionedConcurrencyConfigurationProperty | undefined {
    if (!props.provisionedConcurrentExecutions) {
      return undefined;
    }

    if (props.provisionedConcurrentExecutions <= 0) {
      throw new Error('provisionedConcurrentExecutions must have value greater than or equal to 1');
    }

    return {provisionedConcurrentExecutions: props.provisionedConcurrentExecutions};
  }
}

/**
 * A version/weight pair for routing traffic to Lambda functions
 */
export interface VersionWeight {
  /**
   * The version to route traffic to
   */
  readonly version: IVersion;

  /**
   * How much weight to assign to this version (0..1)
   */
  readonly weight: number;
}
