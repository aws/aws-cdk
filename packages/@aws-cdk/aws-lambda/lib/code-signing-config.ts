import { ISigningProfile } from '@aws-cdk/aws-signer';
import { IResource, Resource, Stack } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnCodeSigningConfig } from './lambda.generated';

/**
 * Code signing configuration policy for deployment validation failure.
 */
export enum UntrustedArtifactOnDeployment {
  /**
   * Lambda blocks the deployment request if signature validation checks fail.
   */
  ENFORCE = 'Enforce',

  /**
   * Lambda allows the deployment of the code package, but issues a warning.
   * Lambda issues a new Amazon CloudWatch metric, called a signature validation error and also stores the warning in CloudTrail.
   */
  WARN = 'Warn',
}

/**
 * A Code Signing Config
 */
export interface ICodeSigningConfig extends IResource {
  /**
   * The ARN of Code Signing Config
   * @attribute
   */
  readonly codeSigningConfigArn: string;

  /**
   * The id of Code Signing Config
   * @attribute
   */
  readonly codeSigningConfigId: string;
}

/**
 * Construction properties for a Code Signing Config object
 */
export interface CodeSigningConfigProps {
  /**
   * List of signing profiles that defines a
   * trusted user who can sign a code package.
   */
  readonly signingProfiles: ISigningProfile[],

  /**
   * Code signing configuration policy for deployment validation failure.
   * If you set the policy to Enforce, Lambda blocks the deployment request
   * if signature validation checks fail.
   * If you set the policy to Warn, Lambda allows the deployment and
   * creates a CloudWatch log.
   *
   * @default UntrustedArtifactOnDeployment.WARN
   */
  readonly untrustedArtifactOnDeployment?: UntrustedArtifactOnDeployment,

  /**
   * Code signing configuration description.
   *
   * @default - No description.
   */
  readonly description?: string,
}

/**
 * Defines a Code Signing Config.
 *
 * @resource AWS::Lambda::CodeSigningConfig
 */
export class CodeSigningConfig extends Resource implements ICodeSigningConfig {
  /**
   * Creates a Signing Profile construct that represents an external Signing Profile.
   *
   * @param scope The parent creating construct (usually `this`).
   * @param id The construct's name.
   * @param codeSigningConfigArn The ARN of code signing config.
   */
  public static fromCodeSigningConfigArn( scope: Construct, id: string, codeSigningConfigArn: string): ICodeSigningConfig {
    const codeSigningProfileId = Stack.of(scope).parseArn(codeSigningConfigArn).resourceName;
    if (!codeSigningProfileId) {
      throw new Error(`Code signing config ARN must be in the format 'arn:aws:lambda:<region>:<account>:code-signing-config:<codeSigningConfigArn>', got: '${codeSigningConfigArn}'`);
    }
    const assertedCodeSigningProfileId = codeSigningProfileId;
    class Import extends Resource implements ICodeSigningConfig {
      public readonly codeSigningConfigArn = codeSigningConfigArn;
      public readonly codeSigningConfigId = assertedCodeSigningProfileId;

      constructor() {
        super(scope, id);
      }
    }
    return new Import();
  }

  public readonly codeSigningConfigArn: string;
  public readonly codeSigningConfigId: string;

  constructor(scope: Construct, id: string, props: CodeSigningConfigProps) {
    super(scope, id);

    const signingProfileVersionArns = props.signingProfiles.map(signingProfile => {
      return signingProfile.signingProfileVersionArn;
    });

    const resource: CfnCodeSigningConfig = new CfnCodeSigningConfig(this, 'Resource', {
      allowedPublishers: {
        signingProfileVersionArns,
      },
      codeSigningPolicies: {
        untrustedArtifactOnDeployment: props.untrustedArtifactOnDeployment ?? UntrustedArtifactOnDeployment.WARN,
      },
      description: props.description,
    });
    this.codeSigningConfigArn = resource.attrCodeSigningConfigArn;
    this.codeSigningConfigId = resource.attrCodeSigningConfigId;
  }
}
