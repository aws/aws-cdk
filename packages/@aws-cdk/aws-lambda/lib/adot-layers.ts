import { RegionInfo } from '@aws-cdk/region-info';
import { IConstruct } from 'constructs';
import { Architecture } from './architecture';
import { IFunction } from './function-base';
import { Stack } from '../../core/lib/stack';
import { Token } from '../../core/lib/token';
import { FactName } from '../../region-info/lib/fact';

/**
 * The type of ADOT Lambda layer
 */
enum AdotLambdaLayerType {
  /**
   * The Lambda layer for ADOT Java instrumentation library. This layer only auto-instruments AWS
   * SDK libraries.
   */
  JAVA_SDK = 'JAVA_SDK',

  /**
   * The Lambda layer for ADOT Java Auto-Instrumentation Agent. This layer automatically instruments
   * a large number of libraries and frameworks out of the box and has notable impact on startup
   * performance.
   */
  JAVA_AUTO_INSTRUMENTATION = 'JAVA_AUTO_INSTRUMENTATION',

  /**
   * The Lambda layer for ADOT Collector, OpenTelemetry for JavaScript and supported libraries.
   */
  JAVASCRIPT_SDK = 'JAVASCRIPT_SDK',

  /**
   * The Lambda layer for ADOT Collector, OpenTelemetry for Python and supported libraries.
   */
  PYTHON_SDK = 'PYTHON_SDK',

  /**
   * The generic Lambda layer that contains only ADOT Collector, used for manual instrumentation
   * use cases (such as Go or DotNet runtimes).
   */
  GENERIC = 'GENERIC',
}

/**
 * Config returned from `AdotLambdaLayerVersion._bind`
 */
interface AdotLambdaLayerBindConfig {
  /**
   * ARN of the ADOT Lambda layer version
   */
  readonly arn: string;
}

/**
 * Return the ARN of an ADOT Lambda layer given its properties. If the region name is unknown
 * at synthesis time, it will generate a map in the CloudFormation template and perform the look
 * up at deployment time.
 *
 * @param scope the parent Construct that will use the imported layer.
 * @param type the type of the ADOT Lambda layer
 * @param version The version of the ADOT Lambda layer
 * @param architecture the architecture of the Lambda layer ('amd64' or 'arm64')
 */
function getLayerArn(scope: IConstruct, type: string, version: string, architecture: string): string {
  const scopeStack = Stack.of(scope);
  const region = scopeStack.region;

  // Region is defined, look up the arn, or throw an error if the version isn't supported by a region
  if (region !== undefined && !Token.isUnresolved(region)) {
    const arn = RegionInfo.get(region).adotLambdaLayerArn(type, version, architecture);
    if (arn === undefined) {
      throw new Error(
        `Could not find the ARN information for the ADOT Lambda Layer of type ${type} and version ${version} in ${region}`,
      );
    }
    return arn;
  }

  // Otherwise, need to add a mapping to be looked up at deployment time
  return scopeStack.regionalFact(FactName.adotLambdaLayer(type, version, architecture));
}

/**
 * Properties for an ADOT instrumentation in Lambda
 */
export interface AdotInstrumentationConfig {
  /**
   * The ADOT Lambda layer.
   */
  readonly layerVersion: AdotLayerVersion;

  /**
   * The startup script to run, see ADOT documentation to pick the right script for your use case: https://aws-otel.github.io/docs/getting-started/lambda
   */
  readonly execWrapper: AdotLambdaExecWrapper;
}

/**
 * An ADOT Lambda layer version that's specific to a lambda layer type and an architecture.
 */
export abstract class AdotLayerVersion {
  /**
   * The ADOT Lambda layer for Java SDK
   *
   * @param version The version of the Lambda layer to use
   */
  public static fromJavaSdkLayerVersion(version: AdotLambdaLayerJavaSdkVersion): AdotLayerVersion {
    return AdotLayerVersion.fromAdotVersion(version);
  }

  /**
   * The ADOT Lambda layer for Java auto instrumentation
   *
   * @param version The version of the Lambda layer to use
   */
  public static fromJavaAutoInstrumentationLayerVersion(
    version: AdotLambdaLayerJavaAutoInstrumentationVersion,
  ): AdotLayerVersion {
    return AdotLayerVersion.fromAdotVersion(version);
  }

  /**
   * The ADOT Lambda layer for JavaScript SDK
   *
   * @param version The version of the Lambda layer to use
   */
  public static fromJavaScriptSdkLayerVersion(version: AdotLambdaLayerJavaScriptSdkVersion): AdotLayerVersion {
    return AdotLayerVersion.fromAdotVersion(version);
  }

  /**
   * The ADOT Lambda layer for Python SDK
   *
   * @param version The version of the Lambda layer to use
   */
  public static fromPythonSdkLayerVersion(version: AdotLambdaLayerPythonSdkVersion): AdotLayerVersion {
    return AdotLayerVersion.fromAdotVersion(version);
  }

  /**
   * The ADOT Lambda layer for generic use cases
   *
   * @param version The version of the Lambda layer to use
   */
  public static fromGenericLayerVersion(version: AdotLambdaLayerGenericVersion): AdotLayerVersion {
    return AdotLayerVersion.fromAdotVersion(version);
  }

  private static fromAdotVersion(adotVersion: AdotLambdaLayerVersion): AdotLayerVersion {
    return new (class extends AdotLayerVersion {
      _bind(_function: IFunction): AdotLambdaLayerBindConfig {
        return {
          arn: adotVersion.layerArn(_function.stack, _function.architecture),
        };
      }
    })();
  }

  /**
   * Produce a `AdotLambdaLayerBindConfig` instance from this `AdotLayerVersion` instance.
   *
   * @internal
   */
  public abstract _bind(_function: IFunction): AdotLambdaLayerBindConfig;
}

/**
 * The wrapper script to be used for the Lambda function in order to enable auto instrumentation
 * with ADOT.
 */
export enum AdotLambdaExecWrapper {
  /**
   * Wrapping regular Lambda handlers.
   */
  REGULAR_HANDLER = '/opt/otel-handler',

  /**
   * Wrapping regular handlers (implementing RequestHandler) proxied through API Gateway, enabling
   * HTTP context propagation.
   */
  PROXY_HANDLER = '/opt/otel-proxy-handler',

  /**
   * Wrapping streaming handlers (implementing RequestStreamHandler), enabling HTTP context
   * propagation for HTTP requests.
   */
  STREAM_HANDLER = '/opt/otel-stream-handler',
}

abstract class AdotLambdaLayerVersion {
  protected constructor(protected readonly type: AdotLambdaLayerType, protected readonly version: string) {}

  /**
   * The ARN of the Lambda layer
   *
   * @param scope The binding scope. Usually this is the stack where the Lambda layer is bound to
   * @param architecture The architecture of the Lambda layer (either X86_64 or ARM_64)
   */
  public layerArn(scope: IConstruct, architecture: Architecture): string {
    return getLayerArn(scope, this.type, this.version, architecture.name);
  }
}

/**
 * The collection of versions of the ADOT Lambda Layer for Java SDK
 */
export class AdotLambdaLayerJavaSdkVersion extends AdotLambdaLayerVersion {
  /**
   * The latest layer version available in this CDK version. New versions could
   * introduce incompatible changes. Make sure to test them before deploying to production.
   */
  public static readonly LATEST = new AdotLambdaLayerJavaSdkVersion('1.19.0');

  /**
   * Version 1.19.0
   */
  public static readonly V1_19_0 = new AdotLambdaLayerJavaSdkVersion('1.19.0');

  private constructor(protected readonly layerVersion: string) {
    super(AdotLambdaLayerType.JAVA_SDK, layerVersion);
  }
}

/**
 * The collection of versions of the ADOT Lambda Layer for Java auto-instrumentation
 */
export class AdotLambdaLayerJavaAutoInstrumentationVersion extends AdotLambdaLayerVersion {
  /**
   * The latest layer version available in this CDK version. New versions could
   * introduce incompatible changes. Make sure to test them before deploying to production.
   */
  public static readonly LATEST = new AdotLambdaLayerJavaAutoInstrumentationVersion('1.19.2');

  /**
   * Version 1.19.2
   */
  public static readonly V1_19_2 = new AdotLambdaLayerJavaAutoInstrumentationVersion('1.19.2');

  private constructor(protected readonly layerVersion: string) {
    super(AdotLambdaLayerType.JAVA_AUTO_INSTRUMENTATION, layerVersion);
  }
}

/**
 * The collection of versions of the ADOT Lambda Layer for Python SDK
 */
export class AdotLambdaLayerPythonSdkVersion extends AdotLambdaLayerVersion {
  /**
   * The latest layer version available in this CDK version. New versions could
   * introduce incompatible changes. Make sure to test them before deploying to production.
   */
  public static readonly LATEST = new AdotLambdaLayerPythonSdkVersion('1.13.0');

  /**
   * Version 1.13.0
   */
  public static readonly V1_13_0 = new AdotLambdaLayerPythonSdkVersion('1.13.0');

  private constructor(protected readonly layerVersion: string) {
    super(AdotLambdaLayerType.PYTHON_SDK, layerVersion);
  }
}

/**
 * The collection of versions of the ADOT Lambda Layer for JavaScript SDK
 */
export class AdotLambdaLayerJavaScriptSdkVersion extends AdotLambdaLayerVersion {
  /**
   * The latest layer version available in this CDK version. New versions could
   * introduce incompatible changes. Make sure to test them before deploying to production.
   */
  public static readonly LATEST = new AdotLambdaLayerJavaScriptSdkVersion('1.7.0');

  /**
   * Version 1.7.0
   */
  public static readonly V1_7_0 = new AdotLambdaLayerJavaScriptSdkVersion('1.7.0');

  private constructor(protected readonly layerVersion: string) {
    super(AdotLambdaLayerType.JAVASCRIPT_SDK, layerVersion);
  }
}

/**
 * The collection of versions of the ADOT Lambda Layer for generic purpose
 */
export class AdotLambdaLayerGenericVersion extends AdotLambdaLayerVersion {
  /**
   * The latest layer version available in this CDK version. New versions could
   * introduce incompatible changes. Make sure to test them before deploying to production.
   */
  public static readonly LATEST = new AdotLambdaLayerGenericVersion('0.62.1');

  /**
   * Version 0.62.1
   */
  public static readonly V0_62_1 = new AdotLambdaLayerGenericVersion('0.62.1');

  private constructor(protected readonly layerVersion: string) {
    super(AdotLambdaLayerType.GENERIC, layerVersion);
  }
}
