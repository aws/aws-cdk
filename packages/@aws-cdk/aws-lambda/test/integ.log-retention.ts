import * as logs from '@aws-cdk/aws-logs';
import * as cdk from '@aws-cdk/core';
import * as lambda from '../lib';

const app = new cdk.App();

const stack = new cdk.Stack(app, 'aws-cdk-lambda-log-retention');

new lambda.Function(stack, 'OneWeek', {
  code: new lambda.InlineCode('exports.handler = (event) => console.log(JSON.stringify(event));'),
  handler: 'index.handler',
  runtime: lambda.Runtime.NODEJS_10_X,
  logRetention: logs.RetentionDays.ONE_WEEK,
});

new lambda.Function(stack, 'OneMonth', {
  code: new lambda.InlineCode('exports.handler = (event) => console.log(JSON.stringify(event));'),
  handler: 'index.handler',
  runtime: lambda.Runtime.NODEJS_10_X,
  logRetention: logs.RetentionDays.ONE_MONTH,
});

new lambda.Function(stack, 'OneYear', {
  code: new lambda.InlineCode('exports.handler = (event) => console.log(JSON.stringify(event));'),
  handler: 'index.handler',
  runtime: lambda.Runtime.NODEJS_10_X,
  logRetention: logs.RetentionDays.ONE_YEAR,
});

app.synth();
