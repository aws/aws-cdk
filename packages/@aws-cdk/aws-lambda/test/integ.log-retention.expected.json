{
  "Resources": {
    "OneWeekServiceRole05A6F9F8": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "OneWeekFE56F6A4": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = (event) => console.log(JSON.stringify(event));"
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "OneWeekServiceRole05A6F9F8",
            "Arn"
          ]
        },
        "Runtime": "nodejs10.x"
      },
      "DependsOn": [
        "OneWeekServiceRole05A6F9F8"
      ]
    },
    "OneWeekLogRetention8E8911C1": {
      "Type": "Custom::LogRetention",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn"
          ]
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "OneWeekFE56F6A4"
              }
            ]
          ]
        },
        "RetentionInDays": 7
      }
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy"
              ],
              "Effect": "Allow",
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB"
          }
        ]
      }
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters11aa2ce8971716ca7c8d28d472ab5e937131e78e136d0de8f4997fb11c4de847S3Bucket46EF559D"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters11aa2ce8971716ca7c8d28d472ab5e937131e78e136d0de8f4997fb11c4de847S3VersionKey68B7BF84"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters11aa2ce8971716ca7c8d28d472ab5e937131e78e136d0de8f4997fb11c4de847S3VersionKey68B7BF84"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn"
          ]
        },
        "Runtime": "nodejs12.x"
      },
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB"
      ]
    },
    "OneMonthServiceRoleFBD1064F": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "OneMonth64E966BF": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = (event) => console.log(JSON.stringify(event));"
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "OneMonthServiceRoleFBD1064F",
            "Arn"
          ]
        },
        "Runtime": "nodejs10.x"
      },
      "DependsOn": [
        "OneMonthServiceRoleFBD1064F"
      ]
    },
    "OneMonthLogRetention814A40D9": {
      "Type": "Custom::LogRetention",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn"
          ]
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "OneMonth64E966BF"
              }
            ]
          ]
        },
        "RetentionInDays": 30
      }
    },
    "OneYearServiceRole24D47762": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "OneYearA82EBDA9": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = (event) => console.log(JSON.stringify(event));"
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "OneYearServiceRole24D47762",
            "Arn"
          ]
        },
        "Runtime": "nodejs10.x"
      },
      "DependsOn": [
        "OneYearServiceRole24D47762"
      ]
    },
    "OneYearLogRetentionBD83A067": {
      "Type": "Custom::LogRetention",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn"
          ]
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "OneYearA82EBDA9"
              }
            ]
          ]
        },
        "RetentionInDays": 365
      }
    }
  },
  "Parameters": {
    "AssetParameters11aa2ce8971716ca7c8d28d472ab5e937131e78e136d0de8f4997fb11c4de847S3Bucket46EF559D": {
      "Type": "String",
      "Description": "S3 bucket for asset \"11aa2ce8971716ca7c8d28d472ab5e937131e78e136d0de8f4997fb11c4de847\""
    },
    "AssetParameters11aa2ce8971716ca7c8d28d472ab5e937131e78e136d0de8f4997fb11c4de847S3VersionKey68B7BF84": {
      "Type": "String",
      "Description": "S3 key for asset version \"11aa2ce8971716ca7c8d28d472ab5e937131e78e136d0de8f4997fb11c4de847\""
    },
    "AssetParameters11aa2ce8971716ca7c8d28d472ab5e937131e78e136d0de8f4997fb11c4de847ArtifactHash27BA7171": {
      "Type": "String",
      "Description": "Artifact hash for asset \"11aa2ce8971716ca7c8d28d472ab5e937131e78e136d0de8f4997fb11c4de847\""
    }
  }
}
